src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -78,45 +75,15 @@ public static BuildEnvironment Instance
         /// <returns>Build environment.</returns>
         private static BuildEnvironment Initialize()
         {
-            // Get the executable we are running
-            var processNameCommandLine = s_getProcessFromCommandLine();
-            var processNameCurrentProcess = s_getProcessFromRunningProcess();
-            var executingAssembly = s_getExecutingAssmblyPath();
-            var currentDirectory = s_getCurrentDirectory();
-            var appContextBaseDirectory = s_getAppContextBaseDirectory();
-
-
-            // Check if our current process name is in the list of own test runners
-            var runningTests = CheckIfRunningTests(processNameCommandLine, processNameCurrentProcess);
-
-            // Check to see if we're running inside of Visual Studio
-            bool runningInVisualStudio;
-            var visualStudioPath = GetVisualStudioPath(out runningInVisualStudio);
-
             var possibleLocations = new Func<BuildEnvironment>[]
             {
-                // Check explicit %MSBUILD_EXE_PATH% environment variable.
-                () => TryFromEnvironmentVariable(runningTests, runningInVisualStudio, visualStudioPath),
-
-                // See if we're running from MSBuild.exe
-                () => TryFromCurrentProcess(processNameCommandLine, runningTests, runningInVisualStudio, visualStudioPath),
-                () => TryFromCurrentProcess(processNameCurrentProcess, runningTests, runningInVisualStudio, visualStudioPath),
-
-                // Try from our current executing assembly (e.g. path to Microsoft.Build.dll)
-                () => TryFromFolder(Path.GetDirectoryName(executingAssembly), runningTests, runningInVisualStudio, visualStudioPath),
-
-                // Try based on the Visual Studio Root
-                ()=> TryFromVisualStudioRoot(visualStudioPath, runningTests, runningInVisualStudio),
-
-#if !CLR2COMPATIBILITY // Assemblies compiled against anything older than .NET 4.0 won't have a System.AppContext
-                // Try the base directory that the assembly resolver uses to probe for assemblies.
-                // Under certain scenarios the assemblies are loaded from spurious locations like the NuGet package cache
-                // but the toolset files are copied to the app's directory via "contentFiles".
-                () => TryFromFolder(appContextBaseDirectory, runningTests, runningInVisualStudio, visualStudioPath),
-#endif
-
-                // Try from the current directory
-                () => TryFromFolder(currentDirectory, runningTests, runningInVisualStudio, visualStudioPath),
+                TryFromEnvironmentVariable,
+                TryFromVisualStudioProcess,
+                TryFromMsBuildProcess,
+                TryFromMsBuildAssembly,
+                TryFromDevConsole,
+                TryFromSetupApi,
+                TryFromAppContextBaseDirectory
             };
```

Comment:
Would be nice to have an API to provide a custom location, e.g. from a public static string property somewhere that is set by caller before entering MSBuild.